Resolves #9: Add more XDMA channels

niclashedam opened the pull request:
I would like a brief review of this code. It should be much of the footwork for resolving #9. I would like to know if what I've done so far is correct and what the next step is.

Commit:
Adds multiple XDMA channels to Hermes

Resolves #9

iomartin commented on hw/misc/hermes.c:
```diff
@@ -966,6 +980,13 @@ static void hermes_instance_init(Object *obj)
     HermesState *hermes = HERMES(obj);
     memcpy(&hermes->bar2, &bar2_init, sizeof(bar2_init));
+
+    for(size_t i = 0; i < H2C_CHANNELS; i++)
```
nit: the qemu coding style uses braces even for blocks with a single statement.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -864,12 +876,14 @@ static void hermes_bar2_write(void *opaque, hwaddr addr, uint64_t val,
 
     trace_hermes_bar2_write(size, addr, val);
 
+    size_t ch = hw2ch(addr);
```
nit: variables should be declared at the [start of the function](https://github.com/qemu/qemu/blob/master/CODING_STYLE.rst#declarations)

iomartin commented on hw/misc/hermes.c:
```diff
@@ -174,6 +177,10 @@ struct hermes_dma_desc {
     hwaddr nxt_addr;
 };
 
+static inline size_t hw2ch(hwaddr addr){
```
maybe rename this to something like `addr2ch`?

iomartin commented on hw/misc/hermes.c:
```diff
@@ -174,6 +177,10 @@ struct hermes_dma_desc {
 };
 
+static inline size_t hw2ch(hwaddr addr){
+    return (addr & 0x1000) >> 8;
```
That mask doesn't seems right to me, I think it should be `0xF00` (as the channel ID is bits 8:11)

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1019,17 +1040,24 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
-    hermes_bar2_init_dir_reg(hermes, &hermes->bar2.h2c);
-    hermes_bar2_init_dir_reg(hermes, &hermes->bar2.c2h);
+
+    for(size_t i = 0; i < H2C_CHANNELS; i++)
```
nit: don't define the iteration variable on the for loop

niclashedam replied:
Do you want the `i` defined on the line above? Why would that change anything?

iomartin replied:
Not at the line above, but at the beginning of the function. It's the QEMU's coding style.

QEMU also provides a scripts/checkpatch.pl script that can verify for style violations. I'd recommended that you configure your git pre-commit hook to [run it automatically](http://blog.vmsplice.net/2011/03/how-to-automatically-run-checkpatchpl.html).

iomartin commented on hw/misc/hermes.c:
```diff
@@ -321,9 +329,10 @@ static const struct hermes_bar2 bar2_init = {
-        .name = "h2c-dma",
-    },
-    .c2h = {
+        .name = "h2c-dma"
```
I think we need a unique `irq_vector` for each channel. It will allow the driver to know which channel has completed. Tables 81 and 85 of [PG195](https://www.xilinx.com/support/documentation/ip_documentation/xdma/v4_1/pg195-pcie-dma.pdf) contain the following:

> The position of the H2C bits always starts at bit 0. The position of the C2H bits is the index above the last H2C index, and therefore depends on the NUM_H2C_CHNL parameter

Also see Figure 9 there for an example.

iomartin replied:
Also, it'd be nice to have unique names for each channel

iomartin commented on hw/misc/hermes.c:
```diff
@@ -287,19 +294,20 @@ static void hermes_dma_thread(HermesState *hermes,
     HermesState *hermes = opaque;
-    hermes_dma_thread(hermes, &hermes->bar2.h2c);
+    for(size_t i = 0; i < H2C_CHANNELS; i++)
+        hermes_dma_thread(hermes, &hermes->bar2.h2c[i]);
```
We don't want the loop here. `hermes_h2c_dma_thread()` is the entry point for the threads created on `hermes_bar2_init_dir_reg()`, so we already have `H2C_CHANNELS` threads running.

The "problem" then is knowing which `hermes_bar_dir_reg` to use when calling `hermes_dma_thread()`. One solution might be to use the dir_reg as the argument when creating the thread. Then you can use the `container_of` macro to get `hermes`.

iomartin reviewed (commented):
A great start @niclashedam!

See some comments below.

As for next steps, I'd suggest you try running parallel DMAs (using different channels) to ensure everything is working. You can use this [example](https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma#quick-example) as a starting point.

niclashedam commented:
I fixed your suggested changes but a single one.
I do not, however, see the channels in `/dev` yet.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -966,6 +981,20 @@ static void hermes_instance_init(Object *obj)
+
+    for(size_t i = 0; i < H2C_CHANNELS; i++){
+        memcpy(&hermes->bar2.h2c[i], &h2c_init, sizeof(h2c_init));
+        hermes->bar2.h2c[i].irq_vector = 2 * i;
```
The H2C channels should have irq_vector in range `[0, H2C_CHANNELS)` and the C2H channels in `[H2C_CHANNELS, H2C_CHANNELS + C2H_CHANNELS)`. See Figure 9 of PG195.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -977,7 +1006,13 @@ static void hermes_bar2_init_dir_reg(HermesState *hermes,
     qemu_cond_init(&dir->dma_cond);
-    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, hermes,
+
+    struct hermes_dir_pair pair = {
```
We dont' need a new struct (plus this won't work because you're allocating on the stack and by the time the thread tries to use it might be gone).

How about something like this?
```diff
diff --git a/hw/misc/hermes.c b/hw/misc/hermes.c
index 41c1e6b643..3f49c5a7ca 100644
--- a/hw/misc/hermes.c
+++ b/hw/misc/hermes.c
@@ -299,15 +299,21 @@ static void hermes_dma_thread(HermesState *hermes,
 
 static void *hermes_h2c_dma_thread(void *opaque)
 {
-    struct hermes_dir_pair *pair = opaque;
-    hermes_dma_thread(pair->hermes, &pair->hermes->bar2.h2c[pair->dir->ch]);
+    struct hermes_bar2_dir_reg *dir = opaque;
+    struct hermes_bar2 *bar2 = container_of(dir, struct hermes_bar2, h2c[dir->ch]);
+    HermesState *hermes = container_of(bar2, HermesState, bar2);
+
+    hermes_dma_thread(hermes, dir);
     return NULL;
 }
 
 static void *hermes_c2h_dma_thread(void *opaque)
 {
-    struct hermes_dir_pair *pair = opaque;
-    hermes_dma_thread(pair->hermes, &pair->hermes->bar2.c2h[pair->dir->ch]);
+    struct hermes_bar2_dir_reg *dir = opaque;
+    struct hermes_bar2 *bar2 = container_of(dir, struct hermes_bar2, c2h[dir->ch]);
+    HermesState *hermes = container_of(bar2, HermesState, bar2);
+
+    hermes_dma_thread(hermes, dir);
     return NULL;
 }
 
@@ -1007,12 +1013,7 @@ static void hermes_bar2_init_dir_reg(HermesState *hermes,
     qemu_mutex_init(&dir->dma_mutex);
     qemu_cond_init(&dir->dma_cond);
 
-    struct hermes_dir_pair pair = {
-        .hermes = hermes,
-        .dir = dir,
-    };
-
-    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, &pair,
+    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, dir,
                        QEMU_THREAD_JOINABLE);
 }
```

niclashedam replied:
Good catch. Thanks.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -966,6 +981,20 @@ static void hermes_instance_init(Object *obj)
+        memcpy(&hermes->bar2.h2c[i], &h2c_init, sizeof(h2c_init));
+        hermes->bar2.h2c[i].irq_vector = 2 * i;
+        hermes->bar2.h2c[i].ch = i;
+        sprintf(hermes->bar2.h2c[i].name, "h2c-dma-%lu", i);
```
You need to allocate `name` first. Or change it to a large enough array instead of a pointer.

niclashedam replied:
Good catch. I had in my mind that it was a char array.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -142,15 +145,16 @@ struct hermes_bar2_dir_reg {
     void *(*entry)(void*);
-    const char *name;
+    char *name;
+    size_t ch;
```
size_t is excessive. int or unsigned int is more than enough.

niclashedam replied:
I agree. It's habit to use the size_t type when I'm in doubt.
I would say that we can even go as far as use a `unsigned char` for now.

lsgunth commented:
Can we please have a better commit message "Resolves #9" is not a commit message. Describe the changes. A Resolves tag may go at the end.

iomartin reviewed (commented):
See comments below.

I think the reason you're not seeing the channels on /dev is because we also need to set the Channel ID bits of the `H2C/C2H Channel Identifier` registers (see Tables 39 and 58).

niclashedam commented:
@lsgunth I will set a more saying commit message before I'm opening this PR for review :+1:
@iomartin I have implemented your suggested changes and added the precommit hook including fixing the errors and warnings. You state that we need to set the Channel ID bits of the H2C/C2H Channel Identifier registers. Does this mean that we need to make `struct hermes_bar2_cfg_reg cfg;` an array in `struct hermes_bar2` ?

iomartin commented on hw/misc/hermes.c:
```diff
@@ -174,6 +178,15 @@ struct hermes_dma_desc {
     hwaddr nxt_addr;
 };
 
+struct hermes_dir_pair {
```
We can now remove this struct

iomartin commented on hw/misc/hermes.c:
```diff
@@ -321,9 +341,9 @@ static const struct hermes_bar2 bar2_init = {
-        .name = "h2c-dma",
-    },
-    .c2h = {
+};
```
We don't need to initialize irq_vector on these structs, as they will be unique to each channel

iomartin commented on hw/misc/hermes.c:
```diff
@@ -174,6 +178,15 @@ struct hermes_dma_desc {
+    struct hermes_bar2_dir_reg *dir;
+};
+
+static inline size_t addr2ch(hwaddr addr){
```
I'd use int here and in the other instances of the code

iomartin commented on hw/misc/hermes.c:
```diff
@@ -977,7 +1015,8 @@ static void hermes_bar2_init_dir_reg(HermesState *hermes,
     qemu_cond_init(&dir->dma_cond);
-    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, hermes,
+
+    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, dir,
```
`hermes_bar2_init_dir_reg` is no longer using `hermes`, so we can remove it from the parameters list.

iomartin reviewed (commented):
> You state that we need to set the Channel ID bits of the H2C/C2H Channel Identifier registers. Does this mean that we need to make struct hermes_bar2_cfg_reg cfg; an array in struct hermes_bar2 ?

No. I'm referring to the `identifier` member of the `hermes_bar2_engine_reg` struct, which we are setting to `HERMES_ID_CORE | HERMES_ID_H2C_CHNL | HERMES_ID_VER_20164`. Now take a look at bits 11:8 of Table 39:

![image](https://user-images.githubusercontent.com/1399327/100466081-91d35400-308d-11eb-8dc7-2f8257b3a864.png)

We're always setting them to 0, but they need to be different for each channel.

niclashedam commented:
Oh Gotcha @iomartin.
I've added the identifier to each channel, but I still don't see it in `/dev`.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -964,20 +981,38 @@ static void hermes_cleanup_msix(HermesState *hermes)
+        memcpy(&hermes->bar2.c2h[i], &c2h_init, sizeof(c2h_init));
+        hermes->bar2.c2h[i].irq_vector = H2C_CHANNELS + i;
+        hermes->bar2.c2h[i].ch = i;
+        hermes->bar2.h2c[i].channel.identifier |= (((H2C_CHANNELS + i) << 8)
```
s/h2c/c2h

Also, you don't need to add `H2C_CHANNELS` to `i` here.

iomartin reviewed (commented):
I made the changes below and I got the devices on /dev. If you still don't see anything, make sure the xdma driver is loaded.

For your reference, the relevant functions on the xdma driver are `probe_engines` and `probe_for_engine`.

niclashedam commented:
I see the problem now. I thought the docker-compose image was building the current source tree, but it seems to pull it from GitHub. It would make sense to add a feature to build the current source tree.

Anyhow, the channels show up and I am able to use them in the test example. This PR is ready for final review.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -142,15 +145,16 @@ struct hermes_bar2_dir_reg {
     void *(*entry)(void*);
-    const char *name;
+    char name[16];
+    unsigned char ch;
```
I'd use int

iomartin commented on hw/misc/hermes.c:
```diff
@@ -819,15 +834,16 @@ static uint64_t hermes_bar2_msix_pba_write(hwaddr addr)
 
 static uint64_t hermes_bar2_read(void *opaque, hwaddr addr, unsigned size)
 {
+    int ch = addr2ch(addr);
```
For safety, we should check that `ch < H2C_CHANNELS` or `ch < C2H_CHANNELS`

iomartin commented on hw/misc/hermes.c:
```diff
@@ -860,16 +876,17 @@ static uint64_t hermes_bar2_read(void *opaque, hwaddr addr, unsigned size)
 static void hermes_bar2_write(void *opaque, hwaddr addr, uint64_t val,
                 unsigned size)
 {
+    int ch = addr2ch(addr);
```
same thing about checking if `ch` is valid

iomartin commented on hw/misc/hermes.c:
```diff
@@ -964,20 +981,37 @@ static void hermes_cleanup_msix(HermesState *hermes)
 
 static void hermes_instance_init(Object *obj)
 {
+    size_t i;
```
use int

iomartin commented on hw/misc/hermes.c:
```diff
@@ -992,6 +1026,7 @@ static void hermes_bar2_destroy_dir_reg(struct hermes_bar2_dir_reg *dir)
 
 static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
 {
+    size_t i;
```
use int

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1019,17 +1054,29 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
 
 static void pci_hermes_uninit(PCIDevice *pdev)
 {
+    size_t i;
```
int

iomartin reviewed (commented):
I tested this and the 4 channels showed up and worked.

A few comments below and it should be ready to go.

Also, can we move the "Resolves #9" from the commit title to the commit message?

niclashedam commented:
Done 😄

iomartin commented on hw/misc/hermes.c:
```diff
@@ -819,15 +834,22 @@ static uint64_t hermes_bar2_msix_pba_write(hwaddr addr)
     HermesState *hermes = opaque;
     uint64_t val = ~0ULL;
 
+    if (ch >= H2C_CHANNELS || ch >= C2H_CHANNELS) {
```
Although unlikely, it is possible to have `H2C_CHANNELS != C2H_CHANNELS`, so we have to check against only one of them. Probably easier to do it in the switch below. Also, it is not exactly an error, more like just an invalid BAR address. I'd just use `hermes_bar_warn_invalid()`.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -964,20 +993,37 @@ static void hermes_cleanup_msix(HermesState *hermes)
+        hermes->bar2.c2h[i].irq_vector = H2C_CHANNELS + i;
+        hermes->bar2.c2h[i].ch = i;
+        hermes->bar2.c2h[i].channel.identifier |= ((i << 8) & 0xF00);
+        sprintf(hermes->bar2.h2c[i].name, "c2h-dma-%u", i);
```
Ops, only noticed this now. This should be c2h, not h2c

niclashedam replied:
Ah - good catch!

niclashedam commented:
Sent back to you for review!

iomartin commented on hw/misc/hermes.c:
```diff
@@ -819,15 +834,22 @@ static uint64_t hermes_bar2_msix_pba_write(hwaddr addr)
     switch ((addr & 0xFFFF) >> 12) {
     case 0x0:
-        val = hermes_bar2_engine_read(hermes, &hermes->bar2.h2c, addr);
+        if (ch >= H2C_CHANNELS) {
```
this has to be an if-else

iomartin reviewed (approved):
Looks good now, thanks!

iomartin commented:
@niclashedam, can you please [sign your commit](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/signing-commits) before I can merge it?

niclashedam commented:
Signed 🔒 🔑
